test/process/index/IndexerTest.cpp

```diff
@@ -0,0 +1,447 @@
+
+        // Why is this a pointer? Just to get access to the file on the command line. It is written if the
+        // Indexer is delete OR enough data was written. If we do not have the pointer, the file gets written after the
+        // test is finished.
+        delete indexer;
+                CHECK(exists(index));
+    }
+
+    TEST (TEST_CREATE_INDEX_W_STREAMED_DATA) {
+        TestResourcesAndFunctions res(INDEXER_SUITE_TESTS, TEST_CREATE_INDEX_SMALL);
+        path fastq = res.getResource(TEST_FASTQ_SMALL);
+        path concat = res.filePath("test_concat.fastq.gz");
+        path index = res.filePath("test_concat.fastq.gz.fqi");
+        path extractedFastq = res.filePath(TEST_FASTQ_SMALL);
+
+
+        int appendCount = 4;
+        bool result = TestResourcesAndFunctions::createConcatenatedFile(fastq, concat, appendCount);
+                CHECK(result);
+                CHECK(4 * file_size(fastq) == file_size(concat));
+
+        ifstream fastqStream(concat.string());
+        auto indexer = new Indexer(
+                make_shared<StreamSource>(&fastqStream),
+                make_shared<PathSink>(index),
+                BlockDistanceStorageStrategy::from(1),
+                true,
+                false,
+                false,
+                true
+        );
+
+        indexer->createIndex();
+        auto storedHeader = indexer->getStoredHeader();
+        auto storedEntries = indexer->getStoredEntries();
+        auto storedLines = indexer->getStoredLines();
+
+                CHECK(indexer->wasSuccessful());
+                CHECK(8 == storedEntries.size());
+
+        delete indexer;
+                CHECK(exists(index));
+    }
+
+    TEST (testIndexerIntegrationTestWPipedSmallDataset) {
+        TestResourcesAndFunctions res(INDEXER_SUITE_TESTS, "testIndexerIntegrationTestWPipedSmallDataset");
+        path fastq = res.getResource(string(TEST_FASTQ_LARGE));
+        path index = res.filePath("test2.fastq.gz.fqi");
+        ifstream fqStream(fastq);
+        auto runner = new IndexerRunner(
+                StreamSource::from(&fqStream),
+                PathSink::from(index),
+                BlockDistanceStorageStrategy::getDefault(),
+                false, false, false, true);
+                CHECK(runner->run() == 0);
+        delete runner;
+                CHECK(file_size(index) > 0);
+    }
+
+    TEST (TEST_CREATE_INDEX_LARGE) {
+        TestResourcesAndFunctions res(INDEXER_SUITE_TESTS, TEST_CREATE_INDEX_LARGE);
+
+        path fastq = res.getResource(string(TEST_FASTQ_LARGE));
+        path index = res.filePath("test2.fastq.gz.fqi");
+
+        uint blockSize = 4;
+
+        auto *indexer = new Indexer(
+                make_shared<PathSource>(fastq),
+                make_shared<PathSink>(index),
+                BlockDistanceStorageStrategy::from(blockSize),
+                true, false, false, true
+        ); // Tell the indexer to store entries. This is solely a debug feature but it
```

Comment:
The sequence of `true` and `false` is really hard to understand. One needs to know exactly at which position which parameter is. Would it not make sense to have e.g. enums with descriptive value names, which are actual types and thus are impossible to get confused (neither during reading code, nor during writing). The values could still be interpretable as booleans (by casting). E.g. nested in the `Indexer` class.

Another option (which I dislike, because it is so crude) is to use binary constants representing bits. This is often used in C-code. I dislike this, because one has to write bit-comparison code, and the interface is less safe and intuitive (e.g. `fun(CAN_PROCESS | OVERWRITE)`).